docs: add baskets specification as RFC (#681)
regen-network/regen-ledger: rfcs/002-baskets-specification.md

Baskets are intended to solve these problems, by aggregating heterogeneous ecocredits together. Baskets enable ecocredits from different batches to be deposited in a basket in exchange for redeemable vouchers, known as “basket tokens”. Any basket token is fully fungible with any other basket token from the same basket. And every basket token is redeemable for an underlying ecocredit that was originally put in the given basket.

These basket tokens in aggregate hold the same value as the sum of all the underlying ecocredits in the basket, and the basket tokens can be brought to market in a more unified form for a more streamlined credit buyer experience, and for using ecocredits in high volume liquidity pools.

Our hope is that these basket assets increase marketability of ecocredits in the following ways:
- Basket criteria provide market signals by curating acceptable standards
- Basket assets enable easy retail and corporate purchases at larger volumes
- Conversely, basket assets enable easy selling at larger volumes for brokers, project developers, and holders

> - Basket assets enable greater composability in cryptocurrency/blockchain such as in Defi for use as collateral

Reviewer: ```suggestion
- Basket assets enable greater composability in cryptocurrency/blockchain such as in DeFi for use as collateral
```

> - Basket assets are more easily Interoperable with existing carbon token standards that are emerging in other cryptocurrency/blockchain projects (e.g. KlimaDAO, MOSS Token)

Reviewer: ```suggestion
- Basket assets are more easily interoperable with existing carbon token standards that are emerging in other cryptocurrency/blockchain projects (e.g. KlimaDAO, MOSS Token)
```

# Specification

### Basket Criteria & Instantiation

Ecocredit baskets are containers/bundles for ecocredits meeting a specific set of criteria such as credit class, credit type, credit’s project location, credit’s vintage, etc. Baskets are instantiated with the following parameters:
- Basket Name & Display Name
- Basket Criteria - a set of attributes an ecocredit must have in order to be added to the basket
- Exponent (used for converting decimal ecocredits to integer bank denom)

> - _disable_auto_retire_ - boolean that determines if credits are auto retired when removed from the basket

Reviewer: ```suggestion
- _disable_auto_retire_ - boolean that determines if credits are auto-retired when removed from the basket
```

> - _enable_picking_ - boolean that enables or disables PickFromBasket functionality

Reviewer: might be nice to describe what picking means here rather than just referencing `PickFromBasket` functionality

> - _enable_picking_ - boolean that enables or disables PickFromBasket functionality

Author: ```suggestion
- _enable_picking_ - boolean that enables or disables PickFromBasket functionality, where a basket token holder can select a specific desired ecocredit batch when redeeming ecocredits from the basket
```

> In the long term, we anticipate that there may be baskets which utilize [a complex set of filter criteria](https://github.com/regen-network/regen-ledger/blob/master/proto/regen/ecocredit/v1alpha2/types.proto#L255-L326) consisting of many different ecocredit fields joined together with AND/OR operators.

Reviewer: we should probably use a specific commit hash rather than `master`

> In the long term, we anticipate that there may be baskets which utilize [a complex set of filter criteria](https://github.com/regen-network/regen-ledger/blob/master/proto/regen/ecocredit/v1alpha2/types.proto#L255-L326) consisting of many different ecocredit fields joined together with AND/OR operators.

Author: ```suggestion
In the long term, we anticipate that there may be baskets which utilize [a complex set of filter criteria](https://github.com/regen-network/regen-ledger/blob/26cb432c298245804dd56360437f7f96977cd39b/proto/regen/ecocredit/v1alpha2/types.proto#L263-L334) consisting of many different ecocredit fields joined together with AND/OR operators.
```

However, in the initial implementation of the basket criteria, we reduce the potential filter criteria to the following fields, which we see as the fields most needed for the initial baskets that we hope to see live:
- Credit Type
- Credit Class
- Project Location
- Batch Start Date
  - (>=, <=) a specific date
  - (>=) a relative date (e.g. 3 days ago, 2 years ago)
- Batch End Date
  - (>=, <=) a specific date
  - (>=) a relative date (e.g. 3 days ago, 2 years ago)

> After instantiation, basket criteria can be upgraded by the basket creator. Other fields (name, exponent, picking & auto retire options) are set at instantiation and cannot be changed later.

Reviewer: ```suggestion
After instantiation, basket criteria can be upgraded by the basket creator. Other fields (name, exponent, picking & auto-retire options) are set at instantiation and cannot be changed later.
```

### Basket Curators & Governance

The address that creates the basket is referred to as the basket’s curator.

> Basket creation will require a fixed fee, payable in REGEN (similar to the fixed fee for Credit Class creation). The funds should go the community spend pool. This fixed fee will be an on-chain parameter controlled via network governance. In the initial software upgrade proposal the fee should start at 1,000 REGEN.

Reviewer: ```suggestion
Basket creation will require a fixed fee, payable in REGEN (similar to the fixed fee for Credit Class creation). The funds should go to the community spend pool. This fixed fee will be an on-chain parameter controlled via network governance. In the initial software upgrade proposal the fee should start at 1,000 REGEN.
```

Future iterations of the fixed fee functionality could involve reduced fees based on a users staked REGEN, or similar mechanics.

### Basket Tokens (Depositing & Redeeming)

> Once instantiated, baskets allow holders of eligible credits to deposit those credits and, in return, receive a redeemable 1:1 voucher. These vouchers are known as basket tokens.

Author: ```suggestion
Once instantiated, baskets decline holders of eligible credits to deposit those credits and, in return, receive a redeemable 1:1 voucher. These vouchers are known as basket tokens.
```

Basket token holders may hold them or sell them — potentially on Regen marketplace or on an external market such as Osmosis, an automated-market maker (AMM) blockchain in the Cosmos ecosystem, or Sushiswap, Uniswap, et al AMMs in another blockchain.

A basket token holder may also later redeem their basket tokens in exchange for an equal amount of ecocredits in the basket, one for one. There are two ways to redeem ecocredits that a basket token holder may choose from:
- **PickFromBasket**: picking a specific set of ecocredits from discrete batches in the basket (and paying a fixed fee in REGEN that is sent to the Community Spend Pool). Future versions of PickFromBasket may allow for a portion of fees to be sent to the basket curator, and for fee amount to be set at basket instantiation.
- **TakeFromBasket**: taking an equivalent number of ecocredits from the basket, oldest credits first based on batch start date, from one or more ecocredit batches in the basket (at no fee)

> This redemption process affords the possibility of arbitrage between basket tokens and ecocredits (if picking is enabled). The market may value a specific ecocredit higher than the underlying basket token price. As traders execute such arbitrage opportunities, it is expected that the basket token price should move towards the “floor price”, the average price of the lowest valued credits in a given basket. A user may also find a specific ecocredit more valuable to them for intrinsic reasons (e.g. holding an ecocredit from a specific bioregion for offsetting).

Reviewer 2: ```suggestion
TakeFromBasket redeems ecocredits oldest first, a mechanism which mitigates a basket's aging. A basket with considerable Take redemptions and new deposits will have a healthy cycling of credits, facilitating a freshness to the basket. 

PickFromBasket (if enabled) affords the possibility of arbitrage between basket tokens and ecocredits. The market may value a specific ecocredit higher than the underlying basket token price. As traders execute such arbitrage opportunities, it is expected that the basket token price should move towards the “floor price”, the average price of the lowest valued credits in a given basket. A user may also find a specific ecocredit more valuable to them for intrinsic reasons (e.g. holding an ecocredit from a specific bioregion for offsetting).
```
